File: src/demos/hello_orchestrator/src/orchestrator_server.cpp
// Orchestrator that dispatches tasks to specialized action servers

#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>
#include <nlohmann/json.hpp>
#include <hello_orchestrator/action/orchestrator_task.hpp>
#include <hello_orchestrator/action/print_message.hpp>
#include <hello_orchestrator/action/move_to_named.hpp>

using OrchestratorTask = hello_orchestrator::action::OrchestratorTask;
using PrintMessage = hello_orchestrator::action::PrintMessage;
using MoveToNamed = hello_orchestrator::action::MoveToNamed;
using GoalHandleOrchestrator = rclcpp_action::ServerGoalHandle<OrchestratorTask>;

class OrchestratorServer : public rclcpp::Node
{
public:
    OrchestratorServer() : Node("orchestrator_server")
    {
        // Create action clients to specialized servers
        print_client_ = rclcpp_action::create_client<PrintMessage>(this, "print_message");
        move_client_ = rclcpp_action::create_client<MoveToNamed>(this, "move_to_named");

        // Create orchestrator action server
        action_server_ = rclcpp_action::create_server<OrchestratorTask>(
            this,
            "orchestrator_task",
            std::bind(&OrchestratorServer::handle_goal, this, std::placeholders::_1, std::placeholders::_2),
            std::bind(&OrchestratorServer::handle_cancel, this, std::placeholders::_1),
            std::bind(&OrchestratorServer::handle_accepted, this, std::placeholders::_1)
        );

        RCLCPP_INFO(this->get_logger(), "Orchestrator action server started");
        RCLCPP_INFO(this->get_logger(), "Waiting for print_message and move_to_named servers...");
    }

private:
    rclcpp_action::Server<OrchestratorTask>::SharedPtr action_server_;
    rclcpp_action::Client<PrintMessage>::SharedPtr print_client_;
    rclcpp_action::Client<MoveToNamed>::SharedPtr move_client_;

    rclcpp_action::GoalResponse handle_goal(
        const rclcpp_action::GoalUUID& /*uuid*/,
        std::shared_ptr<const OrchestratorTask::Goal> /*goal*/)
    {
        RCLCPP_INFO(this->get_logger(), "Received orchestrator goal");
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }

    rclcpp_action::CancelResponse handle_cancel(
        const std::shared_ptr<GoalHandleOrchestrator> /*goal_handle*/)
    {
        RCLCPP_INFO(this->get_logger(), "Received cancel request");
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    void handle_accepted(const std::shared_ptr<GoalHandleOrchestrator> goal_handle)
    {
        std::thread{[this, goal_handle]() {
            execute(goal_handle);
        }}.detach();
    }

    void execute(const std::shared_ptr<GoalHandleOrchestrator> goal_handle)
    {
        RCLCPP_INFO(this->get_logger(), "Executing orchestrator goal");

        const auto goal = goal_handle->get_goal();
        auto feedback = std::make_shared<OrchestratorTask::Feedback>();
        auto result = std::make_shared<OrchestratorTask::Result>();

        // Parse JSON task
        nlohmann::json task_json;
        try {
            task_json = nlohmann::json::parse(goal->task_json);
        } catch (const nlohmann::json::exception& e) {
            result->success = false;
            result->error_message = std::string("JSON parse error: ") + e.what();
            goal_handle->abort(result);
            RCLCPP_ERROR(this->get_logger(), "%s", result->error_message.c_str());
            return;
        }

        if (!task_json.contains("tasks") || !task_json["tasks"].is_array()) {
            result->success = false;
            result->error_message = "JSON must contain 'tasks' array";
            goal_handle->abort(result);
            RCLCPP_ERROR(this->get_logger(), "%s", result->error_message.c_str());
            return;
        }

        auto tasks = task_json["tasks"];
        feedback->total_steps = tasks.size();

        RCLCPP_INFO(this->get_logger(), "📋 ORCHESTRATOR: Executing %zu tasks", tasks.size());

        // Execute each task
        for (size_t i = 0; i < tasks.size(); ++i) {
            const auto& task = tasks[i];

            if (!task.contains("type")) {
                result->success = false;
                result->error_message = "Task " + std::to_string(i) + " missing 'type' field";
                goal_handle->abort(result);
                RCLCPP_ERROR(this->get_logger(), "%s", result->error_message.c_str());
                return;
            }

            std::string type = task["type"];
            feedback->current_step = i + 1;
            feedback->current_action = type;
            goal_handle->publish_feedback(feedback);

            RCLCPP_INFO(this->get_logger(), "  → Step %zu/%zu: %s",
                        i + 1, tasks.size(), type.c_str());

            // Dispatch to appropriate server
            bool success = false;
            if (type == "print") {
                success = execute_print_task(task);
            } else if (type == "move") {
                success = execute_move_task(task);
            } else {
                result->success = false;
                result->error_message = "Unknown task type: " + type;
                goal_handle->abort(result);
                RCLCPP_ERROR(this->get_logger(), "%s", result->error_message.c_str());
                return;
            }

            if (!success) {
                result->success = false;
                result->error_message = "Task " + std::to_string(i) + " (" + type + ") failed";
                result->completed_steps = i;
                goal_handle->abort(result);
                RCLCPP_ERROR(this->get_logger(), "%s", result->error_message.c_str());
                return;
            }

            result->completed_steps = i + 1;
        }

        // All tasks completed successfully
        result->success = true;
        result->error_message = "";
        goal_handle->succeed(result);

        RCLCPP_INFO(this->get_logger(), "✅ ORCHESTRATOR: All tasks completed successfully");
    }

    bool execute_print_task(const nlohmann::json& task)
    {
        if (!task.contains("message")) {
            RCLCPP_ERROR(this->get_logger(), "Print task missing 'message' field");
            return false;
        }

        if (!print_client_->wait_for_action_server(std::chrono::seconds(5))) {
            RCLCPP_ERROR(this->get_logger(), "Print action server not available");
            return false;
        }

        auto goal = PrintMessage::Goal();
        goal.message = task["message"];

        auto goal_handle = print_client_->async_send_goal(goal).get();
        if (!goal_handle) {
            RCLCPP_ERROR(this->get_logger(), "Print goal rejected");
            return false;
        }

        auto result_future = print_client_->async_get_result(goal_handle);
        if (result_future.wait_for(std::chrono::seconds(10)) != std::future_status::ready) {
            RCLCPP_ERROR(this->get_logger(), "Print action timeout");
            return false;
        }

        auto result = result_future.get();
        return result.code == rclcpp_action::ResultCode::SUCCEEDED && result.result->success;
    }

    bool execute_move_task(const nlohmann::json& task)
    {
        if (!task.contains("target")) {
            RCLCPP_ERROR(this->get_logger(), "Move task missing 'target' field");
            return false;
        }

        if (!move_client_->wait_for_action_server(std::chrono::seconds(5))) {
            RCLCPP_ERROR(this->get_logger(), "Move action server not available");
            return false;
        }

        auto goal = MoveToNamed::Goal();
        goal.target_pose = task["target"];

        auto goal_handle = move_client_->async_send_goal(goal).get();
        if (!goal_handle) {
            RCLCPP_ERROR(this->get_logger(), "Move goal rejected");
            return false;
        }

        auto result_future = move_client_->async_get_result(goal_handle);
        if (result_future.wait_for(std::chrono::seconds(60)) != std::future_status::ready) {
            RCLCPP_ERROR(this->get_logger(), "Move action timeout");
            return false;
        }

        auto result = result_future.get();
        return result.code == rclcpp_action::ResultCode::SUCCEEDED && result.result->success;
    }
};

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<OrchestratorServer>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/demos/hello_orchestrator_py/scripts/orchestrator_server.py
#!/usr/bin/env python3
"""Orchestrator server - dispatches JSON tasks to specialized action servers."""

import json

import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer, ActionClient, GoalResponse, CancelResponse
from rclpy.action.server import ServerGoalHandle

from hello_orchestrator_py_interfaces.action import (
    OrchestratorTask,
    PrintMessage,
    MoveToNamedState,
)


class OrchestratorServer(Node):
    """Orchestrator that dispatches tasks to print/move action servers."""

    def __init__(self):
        super().__init__("orchestrator_server_py")

        self._executing = False
        self._print_client = ActionClient(self, PrintMessage, "print_message_py")
        self._move_client = ActionClient(self, MoveToNamedState, "move_to_named_state_py")

        self._action_server = ActionServer(
            self,
            OrchestratorTask,
            "orchestrator_task_py",
            execute_callback=self._execute_callback,
            goal_callback=self._goal_callback,
            cancel_callback=self._cancel_callback,
        )

        self.get_logger().info("Orchestrator server started on 'orchestrator_task_py'")

    def _goal_callback(self, goal_request) -> GoalResponse:
        if self._executing:
            self.get_logger().warning("Goal rejected: orchestrator busy")
            return GoalResponse.REJECT
        return GoalResponse.ACCEPT

    def _cancel_callback(self, goal_handle: ServerGoalHandle) -> CancelResponse:
        return CancelResponse.ACCEPT

    def _execute_callback(self, goal_handle: ServerGoalHandle):
        self._executing = True
        try:
            return self._execute(goal_handle)
        finally:
            self._executing = False

    def _execute(self, goal_handle: ServerGoalHandle) -> OrchestratorTask.Result:
        result = OrchestratorTask.Result()
        result.success = False
        result.completed_steps = 0
        feedback = OrchestratorTask.Feedback()

        try:
            task_json = json.loads(goal_handle.request.task_json)
        except json.JSONDecodeError as e:
            result.error_message = f"JSON parse error: {e}"
            goal_handle.abort()
            return result

        if "tasks" not in task_json or not isinstance(task_json["tasks"], list):
            result.error_message = "JSON must contain 'tasks' array"
            goal_handle.abort()
            return result

        tasks = task_json["tasks"]
        feedback.total_steps = len(tasks)
        self.get_logger().info(f"Executing {len(tasks)} tasks")

        for i, task in enumerate(tasks):
            if goal_handle.is_cancel_requested:
                result.error_message = "Task was canceled"
                result.completed_steps = i
                goal_handle.canceled()
                return result

            task_type = task.get("type")
            if not task_type:
                result.error_message = f"Task {i} missing 'type' field"
                result.completed_steps = i
                goal_handle.abort()
                return result

            feedback.current_step = i + 1
            feedback.current_action = task_type
            goal_handle.publish_feedback(feedback)
            self.get_logger().info(f"Step {i+1}/{len(tasks)}: {task_type}")

            if task_type == "print":
                success = self._dispatch_print(task)
            elif task_type == "move":
                success = self._dispatch_move(task)
            else:
                result.error_message = f"Unknown task type: {task_type}"
                result.completed_steps = i
                goal_handle.abort()
                return result

            if not success:
                result.error_message = f"Task {i} ({task_type}) failed"
                result.completed_steps = i
                goal_handle.abort()
                return result

            result.completed_steps = i + 1

        result.success = True
        result.error_message = ""
        goal_handle.succeed()
        self.get_logger().info("All tasks completed successfully")
        return result

    def _call_action(self, client, goal, timeout_sec: float = 60.0) -> bool:
        """Generic action client helper."""
        if not client.wait_for_server(timeout_sec=5.0):
            self.get_logger().error("Action server not available")
            return False

        send_future = client.send_goal_async(goal)
        rclpy.spin_until_future_complete(self, send_future, timeout_sec=10.0)

        if not send_future.done():
            self.get_logger().error("Goal acceptance timeout")
            return False

        goal_handle = send_future.result()
        if not goal_handle or not goal_handle.accepted:
            self.get_logger().error("Goal rejected")
            return False

        result_future = goal_handle.get_result_async()
        rclpy.spin_until_future_complete(self, result_future, timeout_sec=timeout_sec)

        if not result_future.done():
            self.get_logger().error("Action timeout")
            return False

        return result_future.result().result.success

    def _dispatch_print(self, task: dict) -> bool:
        if "message" not in task:
            self.get_logger().error("Print task missing 'message' field")
            return False
        goal = PrintMessage.Goal()
        goal.message = task["message"]
        return self._call_action(self._print_client, goal, timeout_sec=10.0)

    def _dispatch_move(self, task: dict) -> bool:
        if "target" not in task:
            self.get_logger().error("Move task missing 'target' field")
            return False
        goal = MoveToNamedState.Goal()
        goal.target_pose = task["target"]
        return self._call_action(self._move_client, goal, timeout_sec=60.0)


def main(args=None):
    rclpy.init(args=args)
    node = OrchestratorServer()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
